### DISCUSSION TITLE
Getting "adding a plonk constraint with qM set but xa or xb == 0 (wire 0)" . Should I be worried?

### DISCUSSION THREAD

**Discussion by Scratch-net** [Consensys/gnark#1171] [Q&A] [answered]
I was testing the new version of my circuit and this message popped up. Is this smth I should debug and fix? 
This happened when I called `api.Xor()` which I do a lot in my code
Log is as follows:
```
=== RUN   TestCipher/bn254/plonk
21:43:51 INF compiling circuit
21:43:51 INF parsed circuit inputs nbPublic=1056 nbSecret=352
21:43:51 WRN adding a plonk constraint with qM set but xa or xb == 0 (wire 0)
21:43:51 INF building constraint builder nbConstraints=45351
```

**Comment by ivokub** [comment, marked as answer]
> I was testing the new version of my circuit and this message popped up. Is this smth I should debug and fix? This happened when I called `api.Xor()` which I do a lot in my code Log is as follows:
> 
> ```
> === RUN   TestCipher/bn254/plonk
> 21:43:51 INF compiling circuit
> 21:43:51 INF parsed circuit inputs nbPublic=1056 nbSecret=352
> 21:43:51 WRN adding a plonk constraint with qM set but xa or xb == 0 (wire 0)
> 21:43:51 INF building constraint builder nbConstraints=45351
> ```

Hi, you shouldn't be concerned about the warning, gnark is currently being overly sceptical.

The issue is that in some cases we can deduce that some variable is 0 a la:
```go
...
y := api.Sub(x, x)
z := api.Mul(w, y)
...
```
where `y` is always `0` and thus also `z` is always zero. This warning indicates that in the call to `api.Mul` we have an input which is always 0 and can be optimized out.

Currently we have taken a conservative approach in gnark and do not automatically omit such constraints, so it is left up to the user.

**Reply by Scratch-net** [reply]
Ok cool, thanks
